peeringmod.py

```diff
@@ -0,0 +1,233 @@
+import json
+import os
+import subprocess
+import sys
+from collections import defaultdict
+from dataclasses import asdict, dataclass
+from ipaddress import IPv4Network, IPv6Network
+from typing import List
+from wsgiref import headers
+
+import jinja2
+import jsonschema
+import requests
+
+
+@dataclass
+class AnnouncementGenerator:
+    prefix: IPv4Network or IPv6Network
+    muxes: List[str]
+    details: List[dict]
+
+    def __init__(self, prefix, muxes, **kwargs):
+        self.prefix = prefix
+        self.muxes = muxes
+        self.details = list()
+
+        self.add_details(muxes, **kwargs)
+
+    def add_details(self, muxes, **kwargs):
+        dictdetails = dict()
+
+        dictdetails["muxes"] = muxes
+        for key, value in kwargs.items():
+            dictdetails[key] = value
+
+        self.details.append(dictdetails)
+
+    def create_announcement(self) -> defaultdict:
+        announcement = defaultdict(dict)
+        announcement[self.prefix]["announce"] = self.details
+        return announcement
+
+    def create_withdrawal(self) -> defaultdict:
+        announcement = defaultdict(dict)
+        announcement[self.prefix]["withdraw"] = self.muxes
+        return announcement
```

Comment:
Alright, so it seems we're keeping the announcement around as a dictionary.  Should we instead create something like

```
@dataclasses.dataclass
class PrefixAnnouncement:
    prefix: IPv4Network | IPv6Network
    withdrawals: list[Mux]
    announcements: list[Announcement]

    def validate(schema_fn: pathlib.Path) -> bool:
        ...
```

Reply:
Additionally, we could also look into avoiding the need for `validate` altogether.  This combines two ideas:

1. Only allow the creation of valid objects by carefully checking any created object is valid *on construction*.  For example, we could declare an `enum Mux` and initialize the `enum` dynamically at runtime by reading the valid muxes from the JSON Schema (see [this link][so-dynamic-enum]).  This way people can only initialize instances of `Mux` that are valid.

[so-dynamic-enum]: https://stackoverflow.com/questions/43096541/a-more-pythonic-way-to-define-an-enum-with-dynamic-members

2. Validate instances read from files when reading (we can load the JSON, validate, and *then* initialize the classes from the JSON).  The idea is that users don't need to explicitly call `validate`.